[AArch64] Replace AND with LSL#2 for LDR target (#34101)

ParkHanbum opened the pull request:
Currently, process of replacing bitwise operations consisting of
`LSR`/`LSL` with `And` is performed by `DAGCombiner`.

However, in certain cases, the `AND` generated by this process
can be removed.

Consider following case:
```
        lsr x8, x8, #56
        and x8, x8, #0xfc
        ldr w0, [x2, x8]
        ret
```

In this case, we can remove the `AND` by changing the target of `LDR`
to `[X2, X8, LSL #2]` and right-shifting amount change to 56 to 58.

after changed:
```
        lsr x8, x8, #58
        ldr w0, [x2, x8, lsl #2]
        ret
```

This patch checks to see if the `SHIFTING` + `AND` operation on load
target can be optimized and optimizes it if it can.

davemgreen commented:
Could this be handled better in DAG combine? Maybe with something similar to reassociationCanBreakAddressingModePattern?

ParkHanbum commented:
ok. I'll find way

ParkHanbum committed:
[AArch64] prevent (shl (srl x, c1), c2) -> (and (shift x, c3)) when load

Currently, process of replacing bitwise operations consisting of
`(shl (srl x, c1), c2)` with `And` is performed by `DAGCombiner`.

However, in certain case like `(shl (srl, x, c1) 2)` is do not
need to transform to `AND` if it was used to `Load` Target.

Consider following case:
```
        lsr x8, x8, #56
        and x8, x8, #0xfc
        ldr w0, [x2, x8]
        ret
```

In this case, we can remove the `AND` by changing the target of `LDR`
to `[X2, X8, LSL #2]` and right-shifting amount change to 56 to 58.

after changed:
```
        lsr x8, x8, #58
        ldr w0, [x2, x8, lsl #2]
        ret
```

This patch checks to see if the `(shl (srl x, c1) 2)` operation on
`load` target can be prevent transform to `And`.

davemgreen commented:
Hi. There is code in the DAG combiner to attempt to stop reassociations that would break the addressing mode.
https://github.com/llvm/llvm-project/blob/654846560c8892b32ea1e92c1fffc5e70bf0e75f/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2690
I've not looked into this case in a lot of details, there is a chance the transform happens in a few different places, but could the same thing be done here?

ParkHanbum commented:
@davemgreen Thanks, I think I'm almost there thanks to your advice, I was thinking I should do it in `LOAD`. it's my mistake.

Can I ask one more thing?

I can check the legality of the address via `TLI.isLegalAddressingMode()`, but what function in the TLI can I utilize to check if `PreIndex` is allowed in architecture?

ParkHanbum committed:
[AArch64] Add tests for prevent (shl (srl x, c1), c2) -> (and (shift x, c3)) when load

ParkHanbum commented:
@davemgreen I think it works done. would you review again please?

justinfargnoli commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -563,6 +563,7 @@ namespace {
     SDValue visitFMULForFMADistributiveCombine(SDNode *N);
 
     SDValue XformToShuffleWithZero(SDNode *N);
+    bool isCanBeLoadedWithLsl(SDNode *N);
```
What does `Lsl` stand for in the function name?

justinfargnoli commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -9893,7 +9894,8 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
     if (N0.getOpcode() == ISD::SRL &&
         (N0.getOperand(1) == N1 || N0.hasOneUse()) &&
-        TLI.shouldFoldConstantShiftPairToMask(N, Level)) {
+        TLI.shouldFoldConstantShiftPairToMask(N, Level) &&
+        !isCanBeLoadedWithLsl(N)) {
```
Why not move the call to `isCanBeLoadedWithLsl()` outside of `SelectionDAG` and into the `aarch64` target?

justinfargnoli commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -28338,6 +28340,50 @@ bool DAGCombiner::findBetterNeighborChains(StoreSDNode *St) {
+      return false;
+
+    auto LoadN = dyn_cast<LoadSDNode>(Load);
+    if (!LoadN)
+      return false;
```
If `Load->getOpcode() == ISD::LOAD`, can `dyn_cast<LoadSDNode>(Load) == nullptr`?

justinfargnoli commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -28338,6 +28340,50 @@ bool DAGCombiner::findBetterNeighborChains(StoreSDNode *St) {
+
+    // check load (ldr x, (add x, (shl (srl x, c1) 2)))
+    SDNode *User = N->use_begin().getUse().getUser();
+    if (!User || User->getOpcode() != ISD::ADD)
```
Nit: You could use [`dyn_cast_or_null`](https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates).

justinfargnoli commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -28338,6 +28340,50 @@ bool DAGCombiner::findBetterNeighborChains(StoreSDNode *St) {
+      return false;
+
+    SDNode *Load = User->use_begin().getUse().getUser();
+    if (!Load || Load->getOpcode() != ISD::LOAD)
```
Nit: If `Load->getOpcode() == ISD::LOAD` implies that `dyn_cast<LoadSDNode>(Load) == nullptr`, then you could use [`dyn_cast_or_null`](https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates).

justinfargnoli reviewed (approved):
LGTM, but please wait for @davemgreen's review.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -16918,6 +16918,24 @@ bool AArch64TargetLowering::shouldFoldConstantShiftPairToMask(
+  // We do not need to fold when this shifting used in specific load case:
+  // (ldr x, (add x, (shl (srl x, c1) 2)))
+  if (N->getOpcode() == ISD::SHL) {
+    auto C2 = dyn_cast_or_null<ConstantSDNode>(N->getOperand(1));
```
dyn_cast_or_null -> dyn_cast

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -16918,6 +16918,24 @@ bool AArch64TargetLowering::shouldFoldConstantShiftPairToMask(
+  // (ldr x, (add x, (shl (srl x, c1) 2)))
+  if (N->getOpcode() == ISD::SHL) {
+    auto C2 = dyn_cast_or_null<ConstantSDNode>(N->getOperand(1));
+    if (C2 && C2->getZExtValue() == 2) {
```
Can we change it from C2->getZExtValue() == 2 -> C2->getZExtValue() <= 3? That should capture the common cases where the shift is free.

davemgreen replied:
Oh - should the shift amount match the loaded type? So if it is loading an i32 like the tests, it should be 2, but if it was loading i16/i64 it would need to be different shift amounts?

davemgreen replied:
Yeah sounds good. A few other types sizes: https://godbolt.org/z/GqPT3KbbT. The shift in the addressing mode needs to match the loaded size.

ParkHanbum replied:
https://godbolt.org/z/d6xK3xvav

add comparing with ARM64-GCC.

ParkHanbum replied:
and it works done!

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -16918,6 +16918,24 @@ bool AArch64TargetLowering::shouldFoldConstantShiftPairToMask(
+    auto C2 = dyn_cast_or_null<ConstantSDNode>(N->getOperand(1));
+    if (C2 && C2->getZExtValue() == 2) {
+      auto ShouldADD =
+          dyn_cast_or_null<SDNode>(N->use_begin().getUse().getUser());
```
I don't believe the _or_null is required. Can the `N->use_begin().getUse().getUser()` just be `*N->use_begin()`? It should also probably check that N->hasOneUse().

Same below for the load.

davemgreen reviewed (commented):
Thanks, this does look like a good place for it.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -16918,6 +16918,23 @@ bool AArch64TargetLowering::shouldFoldConstantShiftPairToMask(
+    if (auto C2 = dyn_cast<ConstantSDNode>(N->getOperand(1))) {
+      unsigned ShlAmt = C2->getZExtValue();
+      auto ShouldADD = *N->use_begin();
+      if (ShlAmt <= 3 && ShouldADD && ShouldADD->getOpcode() == ISD::ADD) {
```
You can probably remove the <= 3 now that ShlAmt is checked below.

Also it is probably worth adding a quick check that ShouldADD has Uses, as they are being accessed below.

ParkHanbum replied:
did I properly?

davemgreen replied:
I don't believe that use_begin will return null if there are no uses - it would try to access an invalid iterator. Maybe add a `&& ShouldADD->hasOneUse()` or `&& !ShouldADD->use_empty()` just to make sure it exists and we don't get any weird crashes.

davemgreen reviewed (approved):
Thanks. LGTM

ParkHanbum commented:
I tested this with test-suite in m1 macbook and here is that results:

```
current main

Failed Tests (1):
  test-suite :: SingleSource/Regression/C/gcc-c-torture/execute/GCC-C-execute-alloca-1.test

```

```
This

Failed Tests (2):
  test-suite :: SingleSource/Regression/C/gcc-c-torture/execute/GCC-C-execute-alloca-1.test
  test-suite :: SingleSource/UnitTests/AArch64/aarch64-acle-fmv-features.test
```

what different result at test aarch64-acle-fmv-features.test between main and this. 

```
 22 bti
======main======
 23 simd
======this======
 23         FAIL
 24 simd
 25 fp
```

davemgreen commented:
Does the test reliably fail with the new compiler? It sounds like it might be unrelated, as there have been a number of changes around FMV lately. The test looks like it fork()'s, so maybe something was going wrong with it? If it is a real failure cause by this patch then you might need to debug a bit to see what might be going wrong.

ParkHanbum commented:
sorry, it was my mistake. I was tested it with my built llvm version. I have tested with `Apple clang version 15.0.0` and it passed.

AtariDreams commented:
Any updates on this?

ParkHanbum commented:
nope

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -16918,6 +16918,23 @@ bool AArch64TargetLowering::shouldFoldConstantShiftPairToMask(
+          ShouldADD->getOpcode() == ISD::ADD && ShouldADD->hasOneUse()) {
+        if (auto ShouldLOAD = dyn_cast<LoadSDNode>(*ShouldADD->use_begin())) {
+          unsigned ByteVT = ShouldLOAD->getMemoryVT().getSizeInBits() / 8;
+          if ((1 << ShlAmt) == ByteVT &&
```
-> `(1ULL << ShlAmt) == ByteVT`

davemgreen reviewed (commented):
Sorry - I think I believed you had commit access. I've given this an extra test and it seems OK. If you can update the suggestion to fix a warning, we can get this in.

davemgreen reviewed (approved):
Thanks (I didn't get notification for the update before)

ParkHanbum commented:
@davemgreen never mind. I'm always thanks for your reviewing.
